### DISCUSSION TITLE
Finally got around to begin adapting as a media player, few questions:

### DISCUSSION THREAD

**Discussion by nednoodlehead** [python-ffmpegio/python-ffmpegio#9] [Q&A]
Repo: https://github.com/nednoodlehead/ffmpegio_adaption
(Code originates from the [audio-example](https://github.com/python-ffmpegio/python-ffmpegio-example-audioplayer) library)

First question: Do we have any easy, built-in way to measure how long a song is / how long it has been playing?
I was looking at using `stream.get_time()`, but it seems to not work as I expected, it seems to be a linear timer that has nothing to do with the actual active stream. If this is not possible, I'm guessing the best way we have to measure the time is to have a timer starting when audio is played, and paused when the audio is paused, which brings me to my second question:

Is it possible to begin Audio at a certain time? Looking at the [\_\_init\_\_.py of ffmpegio](https://github.com/python-ffmpegio/python-ffmpegio/blob/main/src/ffmpegio/__init__.py), where the streams (i think thats the right word) come from, there seems to be no room made for any sort of 'time' or any type of argument (even after reading those docs in the file for the `open` function)

Finally, and I know this isn't exactly about `ffmpegio`, but what is the most reasonable way to play music in terms of functionality similar to spotify (or in my case `punge`), where you play and pause songs at will and it acts.. as expected.

The (simplified way) I had `Punge` work was by having the logic and controlling of the songs handled by a single class instance, and that instance would spawn a special type of thread, one that can be killed. It would re-create the thread each time 'Play' was clicked.

Is this inefficient? or is this as good as playing music will get? Could `ffmpegio` be ran on a single thread without the need for consistently killing and creating a new threads? ( The stream.stop_stream() method seems promising for this ! )

**Comment by tikuma-lsuhsc** [comment]
> Do we have any easy, built-in way to measure how long a song is / how long it has been playing?

No

> Is it possible to begin Audio at a certain time? 

Use the FFmpeg's `-ss` input option. So to use it in `ffmpegio` streams they should be opened like `open('input.mp3','ra',ss_in=23.45)` to start from the 23.45-second mark

> what is the most reasonable way to play music in terms of functionality similar to spotify

This is audio player right? (Spotify is, so must be) Then, the best way IMHO to go about this is to separate media reading and playing.

Have the backend to read music files and cache them in memory in raw samples. Stored them in `io.BytesIO` containers as a part of a "song" dict item (along with metadata). Even better, you can get fancier, create a custom class based on it to specify the read arguments in samples instead of bytes (similar to how my streams work).

The backend can expose the `io.BytesIO` object of the current song to the frontend, so it can then read specific data block needed to feed the sound device. Because `io.BytesIO` is seekable, you can seek much easier than relying on FFmpeg.

With this design, `ffmpegio` would be only used to help the backend. If I were you, I'd learn how `ffmpegio` calls FFmpeg and make your own `subprocess.Popen`/`run` call to avoid the unnecessary overhead in `ffmpegio`. This package is really designed to be used in R&D stage and not for a final product.
